### ISSUE TITLE
VCPKG from 7.7.4 to 7.8.1

### ISSUE THREAD

**Issue by krawq** [jtv/libpqxx#743]
Hello,

Quick post in case someone encounter same issue:

I'm using vcpkg, as the package manager recently upgraded the lib to 7.8.1, when upgrading I noticed the following , using latest MSVC.
The package manager doesnt enforce a CXX Standard, so when grabbing the lib , in the CMakeList it uses this:
```
if(NOT "${CMAKE_CXX_STANDARD}")
    set(CMAKE_CXX_STANDARD 17)
endif()
```
So currently i have linker error.
My project is built using c++ 23 but as the package manager used c++17 i have issue with source_location. (using manifest mode so all of that is automatic).

If people encounter this issue. currently override vcpkg.cmake toolchain file with :
```
set(CMAKE_CXX_STANDARD 23)  # Change 20 or 23 as needed
set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
or just stick with 7.7.4 until c++ 20 is enforced in 7.8.x releases.

**Comment by alexander-smyslov** [comment]
This issue is not connected with vcpkg repository:

libpqxx set CMAKE_CXX_STANDARD  by default to 17
https://github.com/jtv/libpqxx/blob/75868467f6808cd712144ebfa2ba26ebbf05e4cb/CMakeLists.txt#L12

if  use this library in project with enable C++20 we get : 
Unable resole  due source_location exist  in C++20 and we have different  signature.
 Error	LNK2019	unresolved external symbol "public:
 __cdecl pqxx::conversion_overrun::conversion_overrun
 (class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,
 struct std::source_location)" (??0
 conversion_overrun@pqxx@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@U
 source_location@3@@Z) referenced in function 
 "public: static char * __cdecl pqxx::string_traits<char const *>
 ::into_buf(char *,char *,char const * const &)" 

Can be fixed in this way (never use pqxx_have_source_location ) 
#if defined(__cpp_lib_source_location) && __cpp_lib_source_location
#  define pqxx_have_source_location 0
#else
#  define pqxx_have_source_location 0
#endif // __cpp_lib_source_location

**Comment by tt4g** [comment]
No. Set `CMAKE_CXX_STANDARD` to `17` only if `CMAKE_CXX_STANDARD` is not set.

https://github.com/jtv/libpqxx/blob/75868467f6808cd712144ebfa2ba26ebbf05e4cb/CMakeLists.txt#L12-L14

libpqxx requires C++17 or higher, so this is a measure for compilers whose default C++ standard version is older.

`CMAKE_CXX_STANDARD` is the CMake variable and can be set from the command line (e.g: `cmake -DCMAKE_CXX_STANDARD=20 ,,,`).
Not setting it is a vcpkg choice, not a libpqxx bug.

**Comment by tt4g** [comment]
Once the PR #740 is merged, there should be no remaining STL symbols that are not supported by the old C++ standard.
However, application binary compatibility is not guaranteed for different versions of the C++ standard.
There is no guarantee that a C++23 project will be able to link libraries built with C++17 and not crash at runtime.

**Comment by tt4g** [comment]
@alexander-smyslov How do you think this problem can be solved?
I believe the correct way is for vcpkg to set `CMAKE_CXX_STANDARD` since vcpkg uses CMake to build libpqxx.
If vcpkg does not set `CMAKE_CXX_STANDARD`, then you should request in the vcpkg repository that they allow you to do that.

**Comment by alexander-smyslov** [comment]
@tt4g as variant - make libpqxx header only library.

**Comment by tt4g** [comment]
@alexander-smyslov Since vcpkg does not support CMake configuration changes, it is too much of a leap to say that the library should be modified into a header-only library.
If you find libraries other than libpqxx that fail to build when different versions of the C++ standard are used, are you going to make them all change to header-only libraries?
I don't know why you haven't made a request to vcpkg to add the ability to change the CMake configuration.

**Comment by tt4g** [comment]
@krawq That is if all the compilation options are the same.
For example, you cannot link with libraries that have different `/MD[d]`, `/MT[d]`, or `/LD[d]` options.
Unlike the C++ standard option flags, this is not a problem that libraries can work around.
Is vcpkg an unfriendly package manager that doesn't solve these problems caused by differences in compiler flags?

**Comment by tt4g** [comment]
@krawq Are the triplet options `x64-windows` and `x64-windows-static` vcpkg options?
If so, then vpckg is making those options configurable because they are important for compatibility, right?
So, if vcpkg implements a feature that allows the C++ standard version of CMake to be configurable in the same way as the triplet option, this problem should be solved.
Why not request vcpkg to implement that feature?

**Comment by tt4g** [comment]
I'm glad the problem has been resolved.
However, it does not look easy to override the configuration with the overlay.
I understand that you think that you should not open a new issue in the vcpkg repository, but I think you should inform the maintainer of your problem.
If the maintainer is aware of a problem in using vcpkg, a feature may be implemented someday.

**Comment by alexander-smyslov** [comment]
I have several projects in one solution (part of project can be compile c++20 and part still on c++17).  before  7.8.0 - I have one precompiled library. right now look  need 2 library instance . so this is not vcpkg issue.

**Comment by KayEss** [comment]
> part of project can be compile c++20 and part still on c++17

I know that the compiler vendors bend over backwards (to the point of destroying advancement of the language), but it isn't really reasonable to expect that different parts of a single executable compiled with different language versions work together cleanly.

For those who are using cmake, you should use `add_subdirectory` to bring libpqxx into your project, that way you can easily get the right build options.

I think I agree with tt4g that this is really a vcpkg problem -- it should (really it needs to) allow a way to set the compile options of dependencies. This seems like table stakes for a C++ package manager to me.

I'm going to reach out in a few places and try to find out what other projects are doing

**Comment by jbrezina** [comment]
I have the same issue:
`error LNK2001: unresolved external symbol pqxx::conversion_error::conversion_error`
The problem is detection of `std::source_location` via `__cpp_lib_source_location` define and the mixed c++ standards. I build all my code with `c++latest` in VS and I'm using vcpkg as a package manager.
I suppose that all the compiler options are passed down to the cmake via `vcpkg/sripts/toolchains/<platform>.cmake` detection, so there is no need to specify any of them in the library or in the vcpkg portfile (as many of them, unfortunately does).
I would mention one of the issues I found in the vcpkg repository:
[[ceres] Should not set CMAKE_CXX_STANDARD](https://github.com/microsoft/vcpkg/issues/22980)
_Note that the upstream Ceres project already [enforces a minimum C++ standard](https://github.com/ceres-solver/ceres-solver/blob/9367ec9cc9914d7ab9048bbcb88a08305c43326e/internal/ceres/CMakeLists.txt#L270-L280) using a more idiomatic and targeted CMake mechanism._
```cmake
# The ability to specify a minimum language version via cxx_std_[11,14,17]
# requires CMake >= 3.8.  Prior to that we have to specify the compiler features
# we require.
if (CMAKE_VERSION VERSION_LESS 3.8)
  set(REQUIRED_PUBLIC_CXX_FEATURES cxx_alignas cxx_alignof cxx_constexpr)
else()
  # Forward whatever C++ version Ceres was compiled with as our requirement
  # for downstream clients.
  set(REQUIRED_PUBLIC_CXX_FEATURES cxx_std_${CMAKE_CXX_STANDARD})
endif()
target_compile_features(ceres PUBLIC ${REQUIRED_PUBLIC_CXX_FEATURES})
```
Maybe libpqxx can take some inspiration in the mentiond [ceres](https://github.com/ceres-solver/ceres-solver) repository?

**Comment by jtv** [comment]
Thanks for the Ceres link @jbrezina.  I'm trying to figure out how this feature works.  Does CMake have a built-in mechanism for turning the list of language features into a configuration header?  Because what we're dealing with here is not really a matter of _required_ language features, it's more a matter of _compiled-in_ language features.  "Which C++ features were present when libpqxx was compiled?"

Of course an added complication is that CMake is not the only build system I need to support.  But in #740 I think I've figured out a way to generate a similar header in a GNU autotools-based build.

Perhaps I should kick this can a little further down the road by putting `source_location` support under the control of a custom macro.

**Comment by jtv** [comment]
@alexander-smyslov the topic of making libpqxx a header-only library comes up sometimes.  One reason why I haven't done that is that I don't want to expose the underlying libpq as a giant abstraction leak.  I'm old enough to remember big piles of printout labeled "Undocumented DOS."  :-)

Here's hoping Modules will become more usable soon...

**Comment by jtv** [comment]
By the way, all this kind of highlights why I _removed_ the setting of a default C++ version from the autotools build.  To build libpqxx using the `configure` script you must either set the language version, or go with whatever your compiler does.

I think setting a default in a library is a bit like the Dark Side of the Force.  Or the Primrose Path.  Faster, easier, more seductive, but leading ultimately to this ticket.

If we remove the default version and only require a minimum language version, would that play well with CMake?

**Comment by KayEss** [comment]
> If we remove the default version and only require a minimum language version, would that play well with CMake?

Unfortunately no, and nor would it if the package managers were using autotools. The problem here is that the C++ version the package manager uses when it compiles libpqxx isn't the same as the version that these projects are using. It doesn't matter how the versions come about, the problem is that they are different.

**Comment by jtv** [comment]
> > If we remove the default version and only require a minimum language version, would that play well with CMake?
> 
> Unfortunately no, and nor would it if the package managers were using autotools. The problem here is that the C++ version the package manager uses when it compiles libpqxx isn't the same as the version that these projects are using. It doesn't matter how the versions come about, the problem is that they are different.

True.  I was just thinking that the healthiest thing to do might be to force the issue: deal with this problem, or else.

**Comment by KayEss** [comment]
> True. I was just thinking that the healthiest thing to do might be to force the issue: deal with this problem, or else.

Yeah, that would be fine if the package managers made it easy to set the compile options for dependencies, but that doesn't seem to be the case.

It's late here, I'm going to sleep on this and wait to see if I get more responses about vcpkg and conan and write something longer in the morning when I'm awake again :-)

**Comment by jtv** [comment]
> Thanks for the Ceres link @jbrezina. I'm trying to figure out how this feature works. Does CMake have a built-in mechanism for turning the list of language features into a configuration header? Because what we're dealing with here is not really a matter of _required_ language features, it's more a matter of _compiled-in_ language features. "Which C++ features were present when libpqxx was compiled?"
> 
> Of course an added complication is that CMake is not the only build system I need to support. But in #740 I think I've figured out a way to generate a similar header in a GNU autotools-based build.
> 
> Perhaps I should kick this can a little further down the road by putting `source_location` support under the control of a custom macro.

Argh.  Bad news.  The current CMake version [only documents features up to C++14](https://cmake.org/cmake/help/latest/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#prop_gbl:CMAKE_CXX_KNOWN_FEATURES).  I'm not going to hold off on supporting new C++ features until CMake defines its own names for them, and then tell people to upgrade CMake first.  That's a _decade_ of lead time!

**Comment by tt4g** [comment]
I have used the package manager conan and know how to deal with these problems.
conan has the ability to specify the C++ standard version via `compiler.cppstd` option to be applied to all packages to be built.
`compiler.cppstd` is specified as `compiler.cppstd=gnu17`. conan will automatically set `CMAKE_CXX_STANDARD` according to `compiler.cppstd` as well.

Furthermore, conan does not request libraries to change their build method for the convenience of the package manager.
Basically, we adjust the `Makefile` and `CMakeLists.txt` so that they can be built by conan with patch files.

**Comment by KayEss** [comment]
Please note firstly that I'm not speaking for libpqxx, I leave that to jtv. These are my personal comments on this situation....

There are two broad schools of thought when it comes to the C++ ecosystem:
1. ABI stability is sacrosanct
2. You compile your own dependencies

The advantage of ABI stability is that you don't need to rebuild so much when things change and it enables the distribution of binaries. The main disadvantage is that APIs can never change (they can't improve and nor can they be fixed). The C++ ABIs are also very complex and it isn't always clear what is and isn't an ABI break -- to put in place a guarantee to not break ABI means being an expert on each ABI where your library is used. ABI stability is the reason we have lost many improvements to the C++ standard library -- we have poorly performing regex and hash tables because the performance improvements we would like would break ABI.

The advantage with re-compiling all the time is that you get to take advantage all improvements to your dependencies and effort is spent on making things better rather than trying to keep things the same. Broken APIs can be fixed, and performance can be improved, and usability can be improved.

I believe that libpqxx has tried to do ABI stability through "best effort". The project is used on too many platforms and in too many different ways for anybody to be sure that the ABI won't be different on at least one platform that people care about. Because of this, the best thing for you to do is to always build libpqxx yourself as part of your project, especially if you want to keep up to date with the latest releases.

Many tool chains make the second choice quite simple. If you're using cmake then you should really be using something like this:

```cmake
add_subdirectory(libpqxx)
target_compile_features(pqxx PUBLIC cxx_std_20)
```

I have an example project that uses [cmake and compiles its own dependencies](https://github.com/KayEss/pqxx-async/).

The difficulties start if you're forced to use a packaged version of libpqxx that does binary distribution -- this is the case for example with most Linux distribution package managers. I think under those circumstances you need to be using the compile settings that the package manager users. This means whoever runs the package manager controls your build options, not you.

vdpkg and conan fall somewhere between these (and they are different to each other). If you are able to use a standard compile triplet (a triplet is something like `x86_64-linux-gnu` used by the system I'm writing this on) then I understand you will generally be able to take advantage of binary distribution, but it's important to note, that by doing this you're ceding control over your build options. If you want to have control over these options then you should define your own triplet and the package managers will compile your dependencies with the options you need. And as tt4g says (thanks @tt4g), with conan you do get some options to control the build without having to go the trouble of defining a whole triplet.

Personally I think that if you're using vcpkg you really do want to be defining your own triplets because you really want to be using things like TSAN and ASAN in your builds, and you really want to compile as many of your dependencies with those options as you can too.

Finally, if you find yourself in a position where you need to use very different compile options to get different parts of your project to compile, I think you have to accept that however you're doing this is going to be brittle and you'll frequently not be able to update some of your components.

**Comment by tt4g** [comment]
This is my personal opinion, but I believe that libpqxx cannot completely solve this problem.

I still think the biggest problem is that vcpkg does not build the package by applying the compile options of the project.
This issue was opened after noticing that the ABI was broken due to the addition of the `std::source_location` symbol in different C++ standard versions.
If this is the only problem, a few techniques could be added to libpqxx to prevent the ABI from breaking.

However, there are limits to what a library can do to make the ABI unbreakable.
The most famous is probably gcc `libstc++`.
This library contains an implementation of `std::string`, but it is no longer compatible with gcc 4.x versions to comply with the C++11 standard.
GCC 5 provides `_GLIBCXX_USE_CXX11_ABI` macros for compatibility with GCC 4, but these macro definitions must be the same in the project and library or compatibility will be broken.

MSVC also has a macro `_ITERATOR_DEBUG_LEVEL` that can easily break ABI.
There are other macros such as `_DEBUG` and `_UNICODE` that can break the ABI in the same source code.

I believe that the ability to share these compilation options between the project and the library is the role of the IDE or package manager, not the library.
And if this problem occurred using vcpkg, then the discussion should take place in the vcpkg repository.
If we discuss it in the libpqxx repository, the vcpkg maintainers will not be aware of it.
This means that the opportunity for vcpkg to improve is lost.

**Comment by jbrezina** [comment]
Because they changed it to these since Cmake 3.8:
[High level meta features indicating C++ standard support](https://cmake.org/cmake/help/v3.27/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#high-level-meta-features-indicating-c-standard-support)

**Comment by jbrezina** [comment]
> > Thanks for the Ceres link @jbrezina. I'm trying to figure out how this feature works. Does CMake have a built-in mechanism for turning the list of language features into a configuration header? Because what we're dealing with here is not really a matter of _required_ language features, it's more a matter of _compiled-in_ language features. "Which C++ features were present when libpqxx was compiled?"
> > Of course an added complication is that CMake is not the only build system I need to support. But in #740 I think I've figured out a way to generate a similar header in a GNU autotools-based build.
> > Perhaps I should kick this can a little further down the road by putting `source_location` support under the control of a custom macro.
> 
> Argh. Bad news. The current CMake version [only documents features up to C++14](https://cmake.org/cmake/help/latest/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#prop_gbl:CMAKE_CXX_KNOWN_FEATURES). I'm not going to hold off on supporting new C++ features until CMake defines its own names for them, and then tell people to upgrade CMake first. That's a _decade_ of lead time!

Because they changed it to these (_`cxx_std_<number>`_) since Cmake 3.8:
[High level meta features indicating C++ standard support](https://cmake.org/cmake/help/v3.27/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#high-level-meta-features-indicating-c-standard-support)

**Comment by tt4g** [comment]
Look above one line.

https://github.com/jtv/libpqxx/blob/75868467f6808cd712144ebfa2ba26ebbf05e4cb/src/CMakeLists.txt#L29-L32

**Comment by jbrezina** [comment]
@KayEss up until now I thought that vcpkg uses the same compiler options for the libraries as the compiled project. Now I'm really confused by these triplets which I thought are only to set up the basic parameters. So all optimizations, architecture, warnings, sdl checks, etc. are ignored for the dependencies? How can that all work together?

**Comment by jtv** [comment]
> Because they changed it to these (_`cxx_std_<number>`_) since Cmake 3.8: [High level meta features indicating C++ standard support](https://cmake.org/cmake/help/v3.27/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#high-level-meta-features-indicating-c-standard-support)

Ah yes, true.  But that still means I can't support new features for users until they make the full jump to a new C++ version.  But for example the C++23 improvement for multidimensional `operator[]` is so useful and soo simple that I added a long time ago.  (It may not be obvious but this can be a performance improvement.)  Sometimes users may even depend on new features.

And as a poorer example, I'm still not confident that I can safely drop the workaround for missing [`to_chars`](https://en.cppreference.com/w/cpp/utility/to_chars) implementations in "C++20" compilers!

**Comment by KayEss** [comment]
> up until now I thought that vcpkg uses the same compiler options for the libraries as the compiled project

I'm no vcpkg expert, just going by how this was all explained to me by people who are vcpkg and conan experts. They have their own reasons for wanting to do what they do, which is based on a worldview of enforced ABI stability.

**Comment by jbrezina** [comment]
I've used a custom triplet for vcpkg where I explicitly set the cxx version, but I still get the compilation error regarding the source location
```cmake
set(VCPKG_TARGET_ARCHITECTURE x64)
set(VCPKG_CRT_LINKAGE static)
set(VCPKG_LIBRARY_LINKAGE static)

set(CMAKE_CXX_STANDARD 23)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

set(CMAKE_C_STANDARD 17)
set(CMAKE_C_STANDARD_REQUIRED ON)

set(VCPKG_CXX_FLAGS "/nologo /permissive- /GL /W3 /Zc:wchar_t /Zi /Gm- /Ob2 /Zc:inline /fp:precise /D_CONSOLE /DWIN64 /DWIN32_LEAN_AND_MEAN /D_UNICODE /DUNICODE /WX- /Zc:forScope /GR- /Gd /Oi /FC /EHsc /Ot" CACHE STRING "")
set(VCPKG_CXX_FLAGS_RELEASE "/GS- /Gy- /O2 /DNDEBUG /MT" CACHE STRING "")
set(VCPKG_CXX_FLAGS_DEBUG "/GS /Od /D_DEBUG /RTC1 /MTd" CACHE STRING "")

set(VCPKG_C_FLAGS "/nologo /permissive- /GL /W3 /Zc:wchar_t /Zi /Gm- /Ob2 /Zc:inline /fp:precise /D_CONSOLE /DWIN64 /DWIN32_LEAN_AND_MEAN /D_UNICODE /DUNICODE /WX- /Zc:forScope /GR- /Gd /Oi /FC /EHsc /Ot" CACHE STRING "")
set(VCPKG_C_FLAGS_RELEASE "/GS- /Gy- /O2 /DNDEBUG /MT" CACHE STRING "")
set(VCPKG_C_FLAGS_DEBUG "/GS /Od /D_DEBUG /RTC1 /MTd" CACHE STRING "")
```

**Comment by tt4g** [comment]
@jbrezina I do not expect vcpkg experts to be looking at this issue.
I recommend that you ask your question in the vcpkg repository.

**Comment by jbrezina** [comment]
this is the question on the library developers, as it has to somehow override the settings to c++17 even if the cxx standard is already set

**Comment by tt4g** [comment]
@jbrezina C++17 is not enforced if vcpkg correctly sets `CMAKE_CXX_STANDARD`.

https://github.com/jtv/libpqxx/blob/75868467f6808cd712144ebfa2ba26ebbf05e4cb/CMakeLists.txt#L12-L14

**Comment by jbrezina** [comment]
@krawq I see. This is really messed up. Unfortunately I tried also setting the CXX and C FLAGS with /std, but it is overriden during the configuration phase of cmake (I suppose because the CMAKE_CXX_STANDARD is not set). My options are enhanced with the configuration options of the library and/or the toolchain options:
```
Run Build Command(s): ... /std:c++latest ... -std:c++17
...
cl : Command line warning D9025 : overriding '/GR' with '/GR-'
cl : Command line warning D9025 : overriding '/Zi' with '/Z7'
cl : Command line warning D9025 : overriding '/Ob2' with '/Ob0'
cl : Command line warning D9025 : overriding '/std:c++latest' with '/std:c++17'
```
So it is probably necessary to also modify the toolchain cmake script if it is possible to set the CMAKE_CXX_STANDARD there, right?

**Comment by jbrezina** [comment]
I've managed to set the vaule in my custom triplet with this:
```cmake
set(VCPKG_CMAKE_CONFIGURE_OPTIONS "-DCMAKE_CXX_STANDARD=23")
```
now it builds with my options, optimizations and no compilation errors. Thanks a lot for consultations

**Comment by alexander-smyslov** [comment]
> I've managed to set the vaule in my custom triplet with this:
> 
> ```cmake
> set(VCPKG_CMAKE_CONFIGURE_OPTIONS "-DCMAKE_CXX_STANDARD=23")
> ```
> 
> now it builds with my options, optimizations and no compilation errors. Thanks a lot for consultations
custom triplet in vcpkg  is answer.
https://github.com/microsoft/vcpkg/pull/34941#issuecomment-1795330700

**Comment by jtv** [comment]
Thanks for the thumbs up — closing.

**Comment by rislah** [comment]
How is custom triplet the answer when the problem is in the library and not the package manager? Custom triplet is just a workaround around the issue. Perhaps fix the issue? Until then switching to another library.

**Comment by tt4g** [comment]
The cause of this problem is the C++ specification.
ABI is not guaranteed when C++ standard versions differ.

This problem occurs with vcpkg because it does not tell CMake the C++ version of the project.

**Comment by jtv** [comment]
This is correct — C++ just does not guarantee that you can link C++17-compiled code to C++20-compiled code and get a working binary.  That is not something that magically changes when you use another library... unless it's not a C++ API but a C API.

Package management could deal with this, but doesn't — and so as far as I'm concerned _that_ is where the issue is.  All we _can_ do is work around it.

**Comment by jtv** [comment]
Some blockers for the next release have cleared up in recent days, so hoping to release 7.9.0 soon, and hopefully the problem won't happen with that.

However in the grander scale, we're still stuck with the problem that all package managers encourage us to link things together that may or may not actually be compatible, because of different language versions etc.
